### DISCUSSION TITLE
[Question] Compiling a standalone example of QuantLib compiled with XAD configuration

### DISCUSSION THREAD

**Discussion by marcin-rybacki** [auto-differentiation/quantlib-xad#10] [General]
Hello there,

first of all, thank you for the effort of integrating QuantLib with your library and sharing it with the community.

I was trying to compile (on a Ubuntu machine) some standalone piece of code (a simple CATCH unit test) linked to QuantLib compiled with XAD (as instructed here https://auto-differentiation.github.io/examples/quantlib/#4-quantlib-cmake-configuration)

Here's my code:

```c++

#include "catch.hpp"

#include <ql/qldefines.hpp>
#include <ql/types.hpp>
#include <ql/pricingengines/blackformula.hpp>

using namespace QuantLib;

void testBlackFormulaForwardDerivative(
    Option::Type optionType,
    const std::vector<Real> &strikes,
    Real bpvol)
{
    Real forward = 1.0;
    Real tte = 10.0;
    Real stdDev = bpvol * std::sqrt(tte);
    Real discount = 0.95;
    Real bump = 0.0001;
    Real epsilon = 1.e-10;

    for (auto it = strikes.begin(); it != strikes.end(); ++it)
    {
        Real strike = *it;
        Real delta = blackFormulaForwardDerivative(
            optionType, strike, forward, stdDev, discount);
        Real bumped_delta = blackFormulaForwardDerivative(
            optionType, strike, forward + bump, stdDev, discount);

        Real base_premium = blackFormula(
            optionType, strike, forward, stdDev, discount);
        Real bumped_premium = blackFormula(
            optionType, strike, forward + bump, stdDev, discount);
        Real delta_approx = bumped_premium - base_premium;

        REQUIRE((std::max(delta * bump, bumped_delta * bump) + epsilon > delta_approx && delta_approx > std::min(
            delta * bump, bumped_delta * bump) - epsilon));
    }
}

TEST_CASE("Black formula forward derivative test.")
{
    std::vector<Real> strikes{0.1, 0.5, 1.0, 2.0, 3.0};
    const Real vol = 0.1;
    testBlackFormulaForwardDerivative(Option::Call, strikes, vol);
    testBlackFormulaForwardDerivative(Option::Put, strikes, vol);
}


```

I compile it using the following CMake:

```cmake
cmake_minimum_required(VERSION 3.15)
project(quantlib_examples)

set(CMAKE_CXX_STANDARD 14)

add_executable(quantlib_examples)

target_sources(quantlib_examples PRIVATE
    src_cpp/black_model_example.cpp
    src_cpp/main.cpp
)

target_include_directories(quantlib_examples PRIVATE
    ${CMAKE_CURRENT_SOURCE_DIR}/deps  # Location of CATCH headers
)

target_link_libraries(quantlib_examples PRIVATE QuantLib)

```

While for the standard QuantLib installation this code (with this CMake configuration) compliles and runs fine, it gives an error `undefined reference to QuantLib::blackFormulaForwardDerivative(QuantLib::Option::Type, double, double, double, double, double)`
when linking it to QuantLib compiled with XAD configuration. I am guessing that the linker is expecting something more, but I cannot quite figure out what.

Before adding any XAD related code to my example, I wanted to start with something simple (that compiles) and I assumed that this code will, with the usual flags. 

I would appreciate any hints or suggestions, thank you!

**Comment by auto-differentiation-dev** [comment]
Hello Marcin,

Thank you for your question. We've been investigating this issue and it turns out that using XAD-enabled QuantLib from a standalone application is indeed broken at the moment. The required CMake flags and linking settings are not exported into the installed target correctly. We have created issue #11 to track this and hope to get it fixed soon.

In the meantime, as a workaround, you could add the following flags to your code to make it work:

```cmake
find_package(XAD REQUIRED)

...
target_link_libraries(quantlib_examples PRIVATE QuantLib XAD::xad)
target_compile_definitions(quantlib_examples PRIVATE QL_INCLUDE_FIRST=ql/qlxad.hpp)
```

(These are the flags added [here in the build file](https://github.com/auto-differentiation/quantlib-xad/blob/main/ql/CMakeLists.txt#L34C55-L41).)

Please let us know if this works for you.

**Reply by marcin-rybacki** [reply]
Thank you very much for responding so promptly. I used your suggestion and the code compiled, much appreciated!

**Comment by auto-differentiation-dev** [comment]
Hi @marcin-rybacki - the issue you had with the integration has now been fixed with this PR #14 - it should be seamless now, if you want to give it a try.

**Reply by marcin-rybacki** [reply]
Hi, thank you for letting me know. I will give it a try!
